### PR TITLE
Improve performance of statistics

### LINKED ISSUE #3750: ruff check --statistics is very slow
Running `--statistics` on Ruff fixtures.
```console
❯ hyperfine 'cargo run -p ruff_cli -- check --no-cache --statistics --select=ALL . || true'
Benchmark 1: cargo run -p ruff_cli -- check --no-cache --statistics --select=ALL . || true
 ⠇ Current estimate: 28.531 s ██████████████████████████████████████████████████████████████████████████████████████████████████████████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ ETA 00:02:23
❯ hyperfine 'cargo run -p ruff_cli -- check --no-cache --select=ALL . || true'
Benchmark 1: cargo run -p ruff_cli -- check --no-cache --select=ALL . || true
  Time (mean ± σ):     799.9 ms ±   7.0 ms    [User: 4226.7 ms, System: 97.1 ms]
  Range (min … max):   788.4 ms … 809.4 ms    10 runs
```

### PR DESCRIPTION
**Pull request by JonathanPlasse**
- Close #3750

Running with `--statistics` is now faster than without it.

```console
❯ hyperfine 'cargo run -p ruff_cli -- check --no-cache --statistics --select=ALL . || true'
Benchmark 1: cargo run -p ruff_cli -- check --no-cache --statistics --select=ALL . || true
  Time (mean ± σ):     749.5 ms ±  17.1 ms    [User: 4181.3 ms, System: 100.7 ms]
  Range (min … max):   721.1 ms … 774.2 ms    10 runs
 
❯ hyperfine 'cargo run -p ruff_cli -- check --no-cache --select=ALL . || true'
Benchmark 1: cargo run -p ruff_cli -- check --no-cache --select=ALL . || true
  Time (mean ± σ):     792.6 ms ±  16.4 ms    [User: 4219.7 ms, System: 94.8 ms]
  Range (min … max):   772.9 ms … 829.8 ms    10 runs
```

### CODE REVIEW

File: crates/ruff_cli/src/printer.rs [hunk 1]
```diff
@@ -471,41 +471,40 @@ impl Printer {
...
     pub fn write_statistics(&self, diagnostics: &Diagnostics) -> Result<()> {
-        let violations: Vec<Rule> = diagnostics
+        let statistics: Vec<ExpandedStatistics> = diagnostics
             .messages
             .iter()
-            .map(|message| message.kind.rule())
+            .map(|message| {
+                (
+                    message.kind.rule().noqa_code(),
+                    message.kind.body.clone(),
```
Comment (charliermarsh) [inline crates/ruff_cli/src/printer.rs:480]: Is it necessary to clone `body` here, and later on too? Not a huge deal, but it seems wasteful given that we only end up using the first `body` anyway.

Comment (JonathanPlasse) [reply crates/ruff_cli/src/printer.rs:480]: I tried sorting and iterating over `Message` instead, but it is slower by more than 2x.
You can go ahead and try to optimize if you find something better.

Comment (charliermarsh) [reply crates/ruff_cli/src/printer.rs:480]: All good, maybe I'll play around with the ownership but this is clearly a huge improvement!

Comment (charliermarsh) [reply crates/ruff_cli/src/printer.rs]: Do you mind testing the version I just pushed?

**Comment by JonathanPlasse** [comment]
```console
# Before
❯ hyperfine 'cargo run -p ruff_cli -- check --no-cache --statistics --select=ALL . || true'
Benchmark 1: cargo run -p ruff_cli -- check --no-cache --statistics --select=ALL . || true
  Time (mean ± σ):     752.7 ms ±  10.7 ms    [User: 4186.4 ms, System: 100.1 ms]
  Range (min … max):   736.7 ms … 769.4 ms    10 runs

# After
❯ hyperfine 'cargo run -p ruff_cli -- check --no-cache --statistics --select=ALL . || true'
Benchmark 1: cargo run -p ruff_cli -- check --no-cache --statistics --select=ALL . || true
  Time (mean ± σ):     739.4 ms ±  14.6 ms    [User: 4179.7 ms, System: 92.5 ms]
  Range (min … max):   723.9 ms … 763.4 ms    10 runs
```

**Pull request merged**